### PR TITLE
Make cps patternable

### PR DESCRIPTION
**Pull request by eefano**
Little hack, you can change the global inner clock with the cps() control function. Very dangerous but very satisfying. You can bring the cps to 0 and stop the playback, also. Negative numbers work in theory (the pianoroll goes in reverse) but no correct audio is produced.

Practical example:

https://youtu.be/No1dQWJKQNU

### CODE REVIEW

**Comment by felixroos** [comment]
I don't think this is a hack.. this partly fixes https://github.com/tidalcycles/strudel/issues/51

You could also use this like `s("bd hh").cps(saw.slow(4).range(.1,1))` which looks less hacky.
To make this complete, I'd say 2 things are missing:

- the .cpm function needs to use .cps internally, so patterning .cpm also behaves like expected
- setcps and friends also need to work with a pattern

+ when https://github.com/tidalcycles/strudel/pull/1004 is merged, this one probably needs to be updated to use this.setCps..

**Comment by yaxu** [comment]
Yes tidal has this functionality too.

Based on my experience with tidal, there could be problems when the cps changes part way through a frame. This can result in events getting doubled up or missed.

**Comment by yaxu** [comment]
I posted a summary of the situation here: https://github.com/tidalcycles/strudel/issues/51#issuecomment-2478814512

File: packages/core/cyclist.mjs [hunk 1]
```diff
@@ -52,6 +52,10 @@ export class Cyclist {
               // see https://github.com/tidalcycles/strudel/pull/1004
               const deadline = targetTime - phase;
               onTrigger?.(hap, deadline, duration, this.cps, targetTime);
+              if (hap.value.cps !== undefined && this.cps != hap.value.cps) {
+                this.cps = hap.value.cps;
+                this.num_ticks_since_cps_change = 0;
```
Comment (yaxu) [inline packages/core/cyclist.mjs:57]: I suppose this shouldn't be 0 as we're partway through a tick. Perhaps somewhere between -1 and 0 ??

Comment (eefano) [reply packages/core/cyclist.mjs:57]: It's an integer counter...  the tick is the smallest unit of handled time. In a tick you feed a fixed-size audio buffer. You could go down to the single sample but i think performance could be killed.

**Comment by yaxu** [comment]
Ok merged, thanks!

**Pull request merged**